src/stdlib_sparse_kinds.fypp

```diff
@@ -0,0 +1,894 @@
+    type, public, extends(sparse_type) :: CSR_type  
+      integer, allocatable  :: col(:)    !> matrix column pointer
+      integer, allocatable  :: rowptr(:) !> matrix row pointer
+    contains
+      procedure :: malloc => malloc_csr
+    end type
+  
+    #:for k1, t1, s1 in (KINDS_TYPES)
+    type, public, extends(CSR_type) :: CSR_${s1}$
+        ${t1}$, allocatable :: data(:) 
+    contains
+        procedure, non_overridable :: at => at_value_csr_${s1}$
+        procedure, non_overridable :: add_value => add_value_csr_${s1}$
+        procedure, non_overridable :: add_block => add_block_csr_${s1}$
+        generic :: add => add_value, add_block
+    end type
+    #:endfor
+
+    !> CSC: Compressed sparse column
+    type, public, extends(sparse_type) :: CSC_type  
+      integer, allocatable  :: colptr(:) !> matrix column pointer
+      integer, allocatable  :: row(:)    !> matrix row pointer
+    contains
+      procedure :: malloc => malloc_csc
+    end type
+  
+    #:for k1, t1, s1 in (KINDS_TYPES)
+    type, public, extends(CSC_type) :: CSC_${s1}$
+        ${t1}$, allocatable :: data(:) 
+    contains
+        procedure, non_overridable :: at => at_value_csc_${s1}$
+        procedure, non_overridable :: add_value => add_value_csc_${s1}$
+        procedure, non_overridable :: add_block => add_block_csc_${s1}$
+        generic :: add => add_value, add_block
+    end type
+    #:endfor
+  
+    !> Compressed ELLPACK
+    type, public, extends(sparse_type) :: ELL_type 
+      integer               :: K = 0 !> maximum number of nonzeros per row
+      integer, allocatable  :: index(:,:) !> column indices
+    contains
+      procedure :: malloc => malloc_ell
+    end type
+  
+    #:for k1, t1, s1 in (KINDS_TYPES)
+    type, public, extends(ELL_type) :: ELL_${s1}$
+        ${t1}$, allocatable :: data(:,:) 
+    contains
+        procedure, non_overridable :: at => at_value_ell_${s1}$
+        procedure, non_overridable :: add_value => add_value_ell_${s1}$
+        procedure, non_overridable :: add_block => add_block_ell_${s1}$
+        generic :: add => add_value, add_block
+    end type
+    #:endfor
+
+    !> Compressed SELL-C 
+    !> Reference : https://library.eecs.utk.edu/storage/files/ut-eecs-14-727.pdf
+    type, public, extends(sparse_type) :: SELLC_type 
+      integer               :: chunk_size = 8 !>  default chunk size
+      integer, allocatable  :: rowptr(:) !> row pointer
+      integer, allocatable  :: col(:,:)  !> column indices
+    end type
+  
+    #:for k1, t1, s1 in (KINDS_TYPES)
+    type, public, extends(SELLC_type) :: SELLC_${s1}$
+        ${t1}$, allocatable :: data(:,:)
+    contains
+        procedure, non_overridable :: at => at_value_sellc_${s1}$
+        procedure, non_overridable :: add_value => add_value_sellc_${s1}$
+        procedure, non_overridable :: add_block => add_block_sellc_${s1}$
+        generic :: add => add_value, add_block
+    end type
+    #:endfor
+
+    !> Conversion from dense to coo
```

Comment:
I suggest to move all these conversion procedures to its own module.

Reply:
These are just module interfaces, the procedures are defined in their own submodule

Reply:
Could these module interfaces be removed from `stdlib_sparse_kinds`, and just be in their own module (and not submodule)? What is the advantage of having them in a submodule?

Reply:
rolled back on the use of submodules [c8d94a3](https://github.com/fortran-lang/stdlib/pull/760/commits/c8d94a3e2544da0f1d40f21967ac424c751c159a) ... I used the submodules to share the constants, they are now in a separated module.